File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {

        // stack

        stack<char> stk;

        for (auto& ch : s) {

            // open - push stack
            if (ch == '(' || ch == '{' || ch == '[') {
                stk.push(ch);
            }
            // close braces - check top if matching or not
            else if (ch == '}' || ch == ']' || ch == ')') {

                // edge case
                if(stk.empty()){
                    return false;
                }

                char top = stk.top();
                stk.pop();

                if ((top == '(' && ch != ')') || (top == '{' && ch != '}') ||
                    (top == '[' && ch != ']')) {
                    return false;
                }
            }
        }


        return stk.empty();
    }
};

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {

        // we will use Deque -> ArrayDeque -> Double Ended Queue
        // TC : O(N), SC : O(N) -> N number of characters
        Deque<Character> stack = new ArrayDeque<>();
        for (int i = 0; i < s.length(); i++) {
            char ch = s.charAt(i);
            if (ch == '(' || ch == '{' || ch == '[') {
                stack.push(ch);
            } else {
                if (stack.isEmpty())
                    return false;

                char top = stack.peek();
                if ((ch == ')' && top == '(') ||
                        (ch == ']' && top == '[') ||
                        (ch == '}' && top == '{')) {
                    stack.pop();
                } else {
                    return false;
                }
            }

        }

        return stack.isEmpty();

    }
}